### PR TITLE
Auth popup 0/n: Add feature_flag="..." view predicate

### PR DESCRIPTION
**Pull request by seanh**
Add a `feature_flag="my_feature"` view predicate (for use in Pyramid `@view_config(...)`'s) that can be used to toggle a view on or off depending on a feature flag, or to direct requests to one view or another depending on a feature flag.

Usage:

        @view_config(route_name="foo", feature_flag="foo")
        def view_that_will_be_called_when_feature_flag_is_on(request):
            ...
    
        @view_config(route_name="foo")
        def view_that_will_be_called_when_feature_flag_is_off(self):
            ...
    
Also add a test page at `/flags/test/view-predicate`.

To test it you can visit http://localhost:8001/flags/test/view-predicate and you should see that `view_that_doesnt_require_feature_flag()` was called. Then turn on the `"foo"` feature flag and visit t http://localhost:8001/flags/test/view-predicate again and you should see that `view_that_requires_feature_flag()` got called. There are a number of ways to turn on the `"foo"` feature flag, for example using a query param: http://localhost:8001/flags/test/view-predicate?feature_flags.foo=true.

### CODE REVIEW

**Comment by lyzadanger** [comment]
WRT:

```
 @view_config(route_name="foo")
    def view_that_will_be_called_when_feature_flag_is_off(self):
        ...
```

Am I correct in parsing this as:

* This view doesn't care what the value of any feature flags are
* This view would be matched whether the feature flag is on or off, technically, right?

Does this then mean that `view_that_will_be_called_when_feature_flag_is_on` needs to be defined earlier in code (so that it is the first match)?

**Comment by seanh** [comment]
The view without the predicate will match either way, yes, but it's normal for multiple views to match a given request and what Pyramid does is it calls the "most specific" of the matching views, meaning the one with the most predicates. It doesn't matter which is defined earlier in the code and the views can be defined in different modules in completely different parts of the code. The algorithm is: Find all views for the matched route; Sort the views in order of most-predicates-first; Evaluate each view, starting with the one with the most predicates, and stop as soon as all of one view's predicates match the request and context.

https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html#view-configuration-parameters

I think it does choose the first defined view if there are two matching views that both have the same number of predicates but I think this'll work fine. If you're trying to use a feature flag to toggle between two views I think those two views will have exactly the same predicates except one of them has the additional feature flag predicate. That holds for the use cases so far anyway.

**Comment by seanh** [comment]
Ah, turns out you can actually [invert any predicate](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html#inverting-predicate-values) anyway, so if you needed to you could do:

```python
from pyramid.config import not_

@view_config(feature_flag=not_("my_feature"), ...)
def view_that_will_be_only_be_called_when_feature_flag_is_off(request):
    ...
```

File: lms/extensions/feature_flags/views/_predicates.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            for example ``"my_feature"``
+        :type feature_flag: str
+        :arg config: the Pyramid config object
+        :type config: pyramid.config.Configurator
+        """
+        self._feature_flag = feature_flag
+
+    def text(self):
+        """
+        Return a string desribing the behavior of this predicate.
```
[the remark this answers is not in the document]
Comment (seanh) [reply lms/extensions/feature_flags/views/_predicates.py]: Thanks!

**Pull request merged**